# Getting started with GitHub Models - OpenAI SDK and embeddings

## 1. Personal access token

A personal access token is made available in the Codespaces environment in the `GITHUB_TOKEN` environment variable. 

## 2. Install dependencies

In [ ]:
%pip install openai --quiet
%pip install python-dotenv --quiet



## 3. Set environment variables and create the client




In [1]:
import os
import dotenv
from openai import OpenAI

dotenv.load_dotenv()

if not os.getenv("GITHUB_TOKEN"):
    raise ValueError("GITHUB_TOKEN is not set")

os.environ["OPENAI_API_KEY"] = os.getenv("GITHUB_TOKEN")
os.environ["OPENAI_BASE_URL"] = "https://models.inference.ai.azure.com/"

client = OpenAI()


## 4. Embed a string

This is just calling the `embeddings.create` endpoint with a simple prompt.




In [3]:
model_name = "text-embedding-3-small" 

response = client.embeddings.create(
    model=model_name,
    input=["Hello, world!"]
)
embeddings = response.data[0].embedding
print(len(embeddings))
print(embeddings[:10])
print(len(response.data))

1536
[-0.019184619188308716, -0.025279032066464424, -0.0017195191467180848, 0.01884828321635723, -0.033795066177845, -0.01969585195183754, -0.02094702236354351, 0.051580529659986496, -0.03212684020400047, -0.030377890914678574]
1


As you can see, the response from the `text-embedding-3-small` model contains a vector of length 1536, which is the embedding of the input string. Different models have different embedding sizes. Please consult the [model documentation](https://github.com/marketplace/models/) for more information about the embedding model you are using. 

## 5. Embed a list of strings

To save on API calls, you can embed a list of strings in a single call. The response will contain a list of embeddings, one for each input string.


In [4]:
model_name = "text-embedding-3-small" 
inputs = ["Hello, world!", "How are you?", "What's the weather like?"]

response = client.embeddings.create(
    model=model_name,
    input=inputs
)
for data in response.data:
    print(data.embedding[:10])

[-0.019184619188308716, -0.025279032066464424, -0.0017195191467180848, 0.01884828321635723, -0.033795066177845, -0.01969585195183754, -0.02094702236354351, 0.051580529659986496, -0.03212684020400047, -0.030377890914678574]
[0.027574976906180382, -0.026575712487101555, -0.027027759701013565, 0.057814665138721466, -0.0034171321894973516, -0.002841662149876356, 0.031809959560632706, 0.010177046060562134, -0.03414158150553703, -0.04465766251087189]
[-0.030613932758569717, -0.006268044002354145, -0.03854816406965256, -0.008153949864208698, 0.0046720425598323345, 0.018029019236564636, 0.02045811340212822, -0.052683304995298386, 0.009106057696044445, 0.01601494289934635]


## Next Steps

A common use case for embeddings in generative AI is to use them to implement Retrieval Augmented Generation (RAG) systems.

See the cookbook [rag_getting_started](../../../cookbooks/python/llamaindex/rag_getting_started.ipynb) for an example of how to do this using the LLamaIndex framework.

To learn more about what you can do with the GitHub models using the OpenAI Python API, [check out theses cookbooks](../../../cookbooks/python/openai/README.md)
